### PR TITLE
FunctionOperator

### PR DESCRIPTION
**Pull request by vpuri3**
need to do tests

### CODE REVIEW

**Comment by vpuri3** [comment]
@ChrisRackauckas take a look?

File: src/sciml.jl [hunk 1]
```diff
@@ -255,66 +255,168 @@ end
...
-             op, op_adjoint, op_inverse, size, p,
+             op, op_adjoint, op_inverse, op_adjoint_inverse, traits, p, t,
             )
     end
+
 end
 
-Base.size(L::FunctionOperator) = L.size
-Base.adjoint(L::FunctionOperator) = FunctionOperator(L.op_adjoint; op_inverse=L.op)
+function FunctionOperator(op, traits=nothing;
```
Comment (ChrisRackauckas) [inline src/sciml.jl:294]: It seems like this is the same as the other kwargs below?

Comment (vpuri3) [reply src/sciml.jl:294]: it's a named tuple that holds those kwargs.

```julia
# src/sciml.jl
    traits = !(traits isa Nothing) ? traits : (;
                                               opnorm = opnorm,
                                               isreal = isreal,
                                               issymmetric = issymmetric,
                                               ishermitian = ishermitian,
                                               isposdef = isposdef,


                                               isinplace = isinplace,
                                               T = T,
                                               size = size,
                                              )```

Comment (vpuri3) [reply src/sciml.jl:294]: makes it easy to pass things around eg, the definition of adjoint:

```julia
Base.size(L::FunctionOperator) = L.traits.size
function Base.adjoint(L::FunctionOperator{iip,T}) where{iip,T}

    if ishermitian(L) | (isreal(L) & issymmetric(L))
        return L
    end

    op = L.op_adjoint
    op_adjoint = L.op

    op_inverse = L.op_adjoint_inverse
    op_adjoint_inverse = L.op_inverse

    traits = (L.traits[1:end-1]..., size=reverse(size(L)))

    p = L.p
    t = L.t

    FuncitonOperator(op, op_adjoint, op_inverse, op_adjoint_inverse, traits, p, t)
end
```

Comment (ChrisRackauckas) [reply src/sciml.jl:294]: But then why have the other keyword arguments? This construction is redundant, unless it's only intended to be an internal and not used/known by users.

Comment (vpuri3) [reply src/sciml.jl:294]: the latter - need it for internals

**Pull request merged**